### DISCUSSION TITLE
Implement UPSERT on buckets

### DISCUSSION THREAD

**Discussion by vic0824** [ArcadeData/arcadedb#718] [Ideas]
Using buckets is a great way of speeding up the queries and achieving scalability, but currently it is not possible to use a bucket as a target for an UPSERT.
This means that UPSERT has to be implemented as two separate queries, one to find out if the record exists, and the other to update or insert as appropriate.
In a scenario in which buckets are used instead of indexes on types, this requires two full scans of the bucket, which could have a detrimental effect on performance.

**Comment by lvca** [comment]
I like this proposal and it shouldn't be hard to implement. In ArcadeDB indexes are always split one per bucket. At a high level you use a TypeIndex, but under the hood there is one index per bucket, so checking if an entry is present should be trivial.

**Comment by lvca** [comment]
Created a new issue: https://github.com/ArcadeData/arcadedb/issues/719

**Reply by lvca** [reply]
Completed. Added some test cases, it works with both `bucket:<name>` and `bucket:<id>`.

**Comment by vic0824** [comment]
Great! Looking forward to the new release.
